libopenstorage/operator
# PWX-34825: Do not continue with PX installation if PX versions configmap or URL are not present/reachable

**What this PR does / why we need it**: Currently when versions configmap is not present or the URL to fetch it is not reachable, we fall back on default images. This behavior is changed and instead of using default images, the storage cluster installation will fail with an appropriate error.

**Which issue(s) this PR fixes**:
Closes # PWX-34825

**Testing Notes**: Tested changes on anthos cluster of 3 nodes
1) Deploy operator image with the changes
2) Ensure there is no versions manifests
3) Add the following env variable to spec of storage cluster yaml file
```
env:
    - name: PX_RELEASE_MANIFEST_URL
      value: "https://install.portworx.com/invalid/version"
```
4) Apply the modified yaml file
Expected results: The storage cluster does not get installed and enters into degraded state. On describing the stc, the following warning is seen
`  Warning  FailedComponent  6m57s (x13 over 7m10s)  storagecluster-controller  StorageCluster reconciliation paused as versions from px-versions ConfigMap cannot be fetched and URL is unreachable due to: release not found`

Positive Test:
1) Deploy operator with changes 
2) Ensure versions manifest does not exist and the URL "https://edge-install.portworx.com/3.1.0/version" is reachable.
3) Add the env variable to storage cluster yaml file :
```
env:
    - name: PX_RELEASE_MANIFEST_URL
      value: "https://edge-install.portworx.com/3.1.0/version"
```
This step can also be skipped. In that case the default URL will be used.
4) Apply the yaml file. Storage cluster installation will proceed without any issues as the versions will be fetched from the URL
During Uninstall, node wiper pods will continue to use default image if versions manifest cannot be fetched. This is done so that storage cluster should never be stuck in a stage where it cannot be uninstalled. Steps to test this, continue from previous test steps
5) Edit the stc yaml file to the incorrect manifest URL again, add uninstall strategy and save changes
6) Delete the storage cluster and watch the logs of the operator. There will be a log stating:

`level=warning msg="Failed to get release versions as versions from px-versions ConfigMap cannot be fetched and URL is unreachable due to: release not found. Using default NodeWiper image" file="uninstall.go:189""`
